pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1905,7 +1902,33 @@ impl RemoteTimelineClient {
                         .await
                         .map_err(|e| anyhow::anyhow!(e))
                 }
-                unexpected @ UploadOp::Barrier(_) | unexpected @ UploadOp::Shutdown => {
+                UploadOp::Barrier(sender) => {
+                    // Barrier operation needs to wait until all deletions are uploaded.
+                    //
+                    // There is an optimization in the deletion queue client that if generations are enabled for a tenant,
+                    // we do not wait for the deletions to be flushed to the remote storage before we tell the user that
+                    // the deletion is complete. This would cause problems for barrier semantics, as we expect all
+                    // deletion to be persistent when the barrier is processed. For example, when we reload a timeline
+                    // without restarting the pageserver (i.e., offload, detach/attach, detach_ancestor), the generation
```

Comment:
Generation numbers are incremented when we detach/attach.

(offload and detach_ancestor may still be enough to motivate the change though)

Reply:
Well, it's already causing all tests to timeout... 😇

Reply:
>Is it okay to also increase generation number for offload/detach instead? 

It would be an elegant thing to do, but because generation numbers have the lifetime of a tenant (not a timeline), we can't do that easily.

I think we could solve the offload case though, by only offloading when something's remote storage client queue is fully drained and they're up to date with the safekeeper commit_lsn (maybe we already check that?).  Then there's no risk that on their next startup they're going to do any delete-create cycles.

For the detach_ancestor case we could do something similar if explicitly stop the walreceiver, flush+freeze, then upload to remote storage.